go/apps/api/routes/v2_keys_get_key/handler.go

```diff
@@ -203,155 +185,169 @@ func (h *Handler) Handle(ctx context.Context, s *zen.Session) error {
-			)
+	// Role permissions
+	if key.RolePermissions != nil {
+		var rolePermissions []db.PermissionInfo
+		if err := json.Unmarshal(key.RolePermissions.([]byte), &rolePermissions); err != nil {
+			h.Logger.Error("failed to unmarshal role permissions", "error", err)
+		} else {
+			for _, p := range rolePermissions {
+				permissionSlugs[p.Slug] = struct{}{}
+			}
 		}
 	}
 
-	permissionSlugs, err := db.Query.ListPermissionsByKeyID(ctx, h.DB.RO(), db.ListPermissionsByKeyIDParams{
-		KeyID: k.KeyId,
-	})
-	if err != nil {
-		return fault.Wrap(err, fault.Code(codes.App.Internal.UnexpectedError.URN()),
-			fault.Internal("unable to find permissions for key"), fault.Public("Could not load permissions for key."))
-	}
 	if len(permissionSlugs) > 0 {
-		k.Permissions = ptr.P(permissionSlugs)
+		slugs := make([]string, 0, len(permissionSlugs))
+		for slug := range permissionSlugs {
+			slugs = append(slugs, slug)
+		}
+		response.Permissions = ptr.P(slugs)
 	}
+}
 
-	// Get roles for the key
-	roles, err := db.Query.ListRolesByKeyID(ctx, h.DB.RO(), k.KeyId)
-	if err != nil {
-		return fault.Wrap(err, fault.Code(codes.App.Internal.UnexpectedError.URN()),
-			fault.Internal("unable to find roles for key"), fault.Public("Could not load roles for key."))
+func (h *Handler) setRoles(response *openapi.KeyResponseData, key db.FindLiveKeyByIDRow) {
+	if key.Roles == nil {
+		return
+	}
+
+	var roles []db.RoleInfo
+	if err := json.Unmarshal(key.Roles.([]byte), &roles); err != nil {
+		h.Logger.Error("failed to unmarshal roles", "error", err)
+		return
 	}
 
 	if len(roles) > 0 {
 		roleNames := make([]string, len(roles))
 		for i, role := range roles {
 			roleNames[i] = role.Name
 		}
+		response.Roles = ptr.P(roleNames)
+	}
+}
 
-		k.Roles = ptr.P(roleNames)
+func (h *Handler) setRatelimits(response *openapi.KeyResponseData, key db.FindLiveKeyByIDRow) {
+	if key.Ratelimits == nil {
+		return
 	}
 
-	return s.JSON(http.StatusOK, Response{
-		Meta: openapi.Meta{
-			RequestId: s.RequestID(),
-		},
-		Data: k,
-	})
+	var ratelimits []db.RatelimitInfo
+	if err := json.Unmarshal(key.Ratelimits.([]byte), &ratelimits); err != nil {
+		h.Logger.Error("failed to unmarshal ratelimits", "error", err)
+		return
+	}
+
+	var keyRatelimits []openapi.RatelimitResponse
+	var identityRatelimits []openapi.RatelimitResponse
+
+	for _, rl := range ratelimits {
+		ratelimitResp := openapi.RatelimitResponse{
+			Id:        rl.ID,
+			Duration:  rl.Duration,
+			Limit:     int64(rl.Limit),
+			Name:      rl.Name,
+			AutoApply: rl.AutoApply,
+		}
+
+		// Add to key ratelimits if it belongs to this key
+		if rl.KeyID.Valid && rl.KeyID.String == key.ID {
+			keyRatelimits = append(keyRatelimits, ratelimitResp)
+		}
+
+		// Also add to identity ratelimits if it has an identity_id that matches
+		if rl.IdentityID.Valid && key.IdentityID.Valid && rl.IdentityID.String == key.IdentityID.String {
+			identityRatelimits = append(identityRatelimits, ratelimitResp)
+		}
```

Comment:
This will always be the case since we are always searching for ratelimits that are either matching the keyId or the attached identityId